nncf/torch/tensor_statistics/collectors.py

```diff
@@ -108,153 +142,267 @@ def sum(tensor: NNCFTensor) -> TensorElementsType:
-        median_tensor = torch.from_numpy(numpy_median).to(dtype=torch.float)
-        mad_tensor = torch.from_numpy(numpy_mad).to(dtype=torch.float)
+class PTNoopReducer(PTReducerMixIn, NoopReducer):
+    pass
 
-        median_tensor = expand_like(median_tensor, list(self._reduction_shape))
-        mad_tensor = expand_like(mad_tensor, list(self._reduction_shape))
 
-        return PTMedianMADTensorStatistic(median_tensor, mad_tensor)
+class PTMinReducer(PTReducerMixIn, MinReducer):
+    pass
 
 
-class PTPercentileStatisticCollector(PercentileStatisticCollector):
-    def _register_input(self, x: torch.Tensor):
-        with no_nncf_trace():
-            self._samples.append(x.detach().cpu().numpy())
+class PTMaxReducer(PTReducerMixIn, MaxReducer):
+    pass
 
-    def _get_statistics(self) -> PTPercentileTensorStatistic:
-        percentile_vs_values_dict = self._prepare_statistics()
-        for key, val in percentile_vs_values_dict.items():
-            torch_percentiles = torch.from_numpy(val).to(dtype=torch.float)
-            percentile_vs_values_dict[key] = expand_like(torch_percentiles, list(self._reduction_shape))
-        return PTPercentileTensorStatistic(percentile_vs_values_dict)
 
+class PTAbsMaxReducer(PTReducerMixIn, AbsMaxReducer):
+    pass
 
-class PTMeanPercentileStatisticCollector(MeanPercentileStatisticCollector):
-    def _register_input(self, x: torch.Tensor):
-        with no_nncf_trace():
-            for pct, val in self._all_pct_values.items():
-                np_vals = np_percentile_reduce_like(x.cpu().numpy(), self._reduction_shape, pct)
-                torch_vals = torch.from_numpy(np_vals).to(dtype=torch.float)
-                val.append(torch_vals)
 
-    def _get_statistics(self) -> PTPercentileTensorStatistic:
-        mean_percentile_values = {}
-        for pct, val in self._all_pct_values.items():
-            stacked_pct_vals = torch.stack(list(val))
-            mean_percentile_values[pct] = stacked_pct_vals.mean(dim=0).view(self._reduction_shape)
-        return PTPercentileTensorStatistic(mean_percentile_values)
+class PTMeanReducer(PTReducerMixIn, MeanReducer):
+    pass
 
 
-class PTMeanStatisticCollector(MeanStatisticCollector):
-    @staticmethod
-    def _get_processor() -> NNCFCollectorTensorProcessor:
-        return PTNNCFCollectorTensorProcessor()
+class PTQuantileReducer(PTReducerMixIn, QuantileReducer):
+    pass
+
 
-    def _register_input(self, x: torch.Tensor):
-        with no_nncf_trace():
-            self._register_input_common(PTNNCFTensor(x))
+class PTAbsQuantileReducer(PTReducerMixIn, AbsQuantileReducer):
+    pass
+
+
+class PTBatchMeanReducer(PTReducerMixIn, BatchMeanReducer):
+    pass
 
-    def _get_statistics(self) -> PTMeanTensorStatistic:
-        return PTMeanTensorStatistic(self._mean_aggregate().tensor, self._shape())
+
+class PTMeanPerChanelReducer(PTReducerMixIn, MeanPerChReducer):
+    pass
+
+
+def get_min_max_statistic_collector(
```

Comment:
There were no any docstring in any tensor collector, I mean

Reply:
Update: I believe `init_config.init_type` as a string is a root of this issue. I believe we should leave this refactoring for other PR to keep "manageable" size of this one

Reply:
It is not better than the previous implementation. In fact, the previous implementation was an honest factory, the way it is described everywhere ([one](https://www.gofpattern.com/creational/patterns/factoryMethodPattern-java-code.php), [two](https://www.oodesign.com/factory-pattern), [three](https://www.modernescpp.com/index.php/factory-method/)). In its basic implementation, there is no way around `if` or `switch` statements. That Python easily allows to build dictionaries for function arguments does not mean that we should use it. For Python specifically, an average Python programmer doesn't know any better, so we end up with SO answers mainly proposing the `**kwargs` approach. The problem with this is that typing information is lost, and I'm not the only one who thinks that: https://stackoverflow.com/questions/35916604/factory-method-for-objects-with-many-parameters#comment131703112_35917230

When a dev looks at a factory method, it is usually to make sure that the input parameters to the factory method get dispatched correctly to the specific class. With the previous implementation, it is immediately visible how the constructor arguments for each class variation get deduced from the factory input parameters. With current implementation, you have to look through the entire code of the factory, tracking in your mind the current state of the kwargs dict, and even after all that having no clue if the dict was constructed correctly in terms of arg count, name or type. If you want to rename an argument in one of the created classes using IDE, the IDE won't refactor the keys you are building in this dict, and you would have to run tests and see them fail to find that out.

Typing information is beneficial here since this exact factory is not so much of a point of scale, but it is now a point of serious failure. The `**kwargs` approach would have been good if we had hundreds of classes we had to dispatch here, and even then we should have been implementing better typing support via `TypedDict` or `Protocol` as mentioned [here](https://stackoverflow.com/questions/74382168/python-factory-functions-with-different-arguments).

Reply:
Got your point, done